Segfault during debug build on FreeBSD AArch64

I can no longer do a debug build on FreeBSD 14.1 AArch64. It's been a bit since the last time I did that rather than a typical build, so I'm not sure when exactly the regression here occurred.

Relevant output from `gmake debug VERBOSE=1`:
```
 cd /root/julia/base && WINEPATH="/root/julia/usr/bin;$WINEPATH" JULIA_NUM_THREADS=1  /root/julia/usr/bin/julia-debug -O0 -C "native"  --output-o /root/julia/usr/lib/julia/basecompiler-debug-o.a.tmp --startup-file=no --warn-overwrite=yes --depwarn=error -g0 Base_compiler.jl --buildroot ./  --dataroot ../usr/share/
JIT session error: In graph jitted-1, section .text: relocation target "__stack_chk_guard" at address 0x860c93c8 is out of range of Page21 fixup at 0x40b3a8d3d010 (japi1_top-level scope_1, 0x40b3a8d3d000 + 0x10)

[51949] signal 11 (1): Segmentation fault
in expression starting at boot.jl:283
ArrayRef at /root/julia/usr/include/llvm/ADT/ArrayRef.h:74
cantFail<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef> > > at /root/julia/usr/include/llvm/Support/Error.h:785
findSymbols at /root/julia/src/jitlayers.cpp:2154
jl_compile_codeinst_now at /root/julia/src/jitlayers.cpp:725
jl_generate_fptr_for_unspecialized_impl at /root/julia/src/jitlayers.cpp:924
jl_compile_method_internal at /root/julia/src/gf.c:3551
_jl_invoke at /root/julia/src/gf.c:4007
ijl_invoke at /root/julia/src/gf.c:4022
jl_toplevel_eval_flex at /root/julia/src/toplevel.c:762
jl_parse_eval_all at /root/julia/src/toplevel.c:890
ijl_load_ at /root/julia/src/toplevel.c:934
ijl_load at /root/julia/src/toplevel.c:947
_finish_jl_init_ at /root/julia/src/init.c:603
ijl_init_ at /root/julia/src/init.c:783
jl_repl_entrypoint at /root/julia/src/jlapi.c:1131
jl_load_repl at /root/julia/cli/loader_lib.c:601
main at /root/julia/usr/bin/julia-debug (unknown line)
__libc_start1 at /lib/libc.so.7 (unknown line)
Allocations: 0 (Pool: 0; Big: 0); GC: 0
gmake[1]: *** [sysimage.mk:148: /root/julia/usr/lib/julia/basecompiler-debug-o.a] Segmentation fault (core dumped)
gmake[1]: Leaving directory '/root/julia'
gmake: *** [Makefile:114: julia-sysimg-debug] Error 2
```

We're not currently setting `HAVE_SSP=1` on FreeBSD for whatever reason, but we could, though I have no idea whether it's relevant here. We could also import `__stack_chk_guard` from libc like we do on macOS. As a guess, I tried making this change locally and rebuilding:

```diff
diff --git a/Make.inc b/Make.inc
index 049d0bebe0..babf6202dc 100644
--- a/Make.inc
+++ b/Make.inc
@@ -1539,6 +1539,7 @@ JLIBLDFLAGS += -Wl,-z,noexecstack
 endif

 ifeq ($(OS), FreeBSD)
+HAVE_SSP := 1
 JLDFLAGS += -Wl,-Bdynamic
 OSLIBS += -lelf -lkvm -lrt -lpthread -latomic

diff --git a/src/julia_internal.h b/src/julia_internal.h
index b6b77b2421..5846ee7615 100644
--- a/src/julia_internal.h
+++ b/src/julia_internal.h
@@ -176,8 +176,8 @@ static inline void msan_unpoison_string(const volatile char *a) JL_NOTSAFEPOINT
 #endif
 #endif

-#if defined(HAVE_SSP) && defined(_OS_DARWIN_)
-// On Darwin, this is provided by libSystem and imported
+#if defined(HAVE_SSP) && (defined(_OS_DARWIN_) || defined(_OS_FREEBSD_))
+// This is provided by libSystem on Darwin and libc on FreeBSD, and imported
 extern JL_DLLIMPORT uintptr_t __stack_chk_guard;
 #elif defined(HAVE_SSP)
 // Added by compiler runtime in final link - not DLLIMPORT
```

But alas, the result is the same.

To see if I could get any more information, I ran the failing build step under LLDB:

```
* thread #1, name = 'julia-debug', stop reason = signal SIGSEGV: address not mapped to object (fault address: 0x0)
    frame #0: 0x000000008e42571c libjulia-codegen-debug.so.1.13`llvm::ArrayRef<llvm::orc::JITDylib*>::ArrayRef(this=0x0000000000000000, OneElt=0x0000000000000000) at ArrayRef.h:74:11
   71
   72       /// Construct an ArrayRef from a single element.
   73       /*implicit*/ ArrayRef(const T &OneElt LLVM_LIFETIME_BOUND)
-> 74           : Data(&OneElt), Length(1) {}
   75
   76       /// Construct an ArrayRef from a pointer and length.
   77       constexpr /*implicit*/ ArrayRef(const T *data LLVM_LIFETIME_BOUND,
(lldb) thread backtrace
* thread #1, name = 'julia-debug', stop reason = signal SIGSEGV: address not mapped to object (fault address: 0x0)
  * frame #0: 0x000000008e42571c libjulia-codegen-debug.so.1.13`llvm::ArrayRef<llvm::orc::JITDylib*>::ArrayRef(this=0x0000000000000000, OneElt=0x0000000000000000) at ArrayRef.h:74:11
    frame #1: 0x000000008e4256c4 libjulia-codegen-debug.so.1.13`llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>> llvm::cantFail<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef>>>(ValOrErr=Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr, void>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::orc::ExecutorSymbolDef> > > @ 0x00000000806df9d8, Msg="Failure value returned from cantFail wrapped call") at Error.h:785:7
    frame #2: 0x000000008e411f78 libjulia-codegen-debug.so.1.13`JuliaOJIT::findSymbols(this=0x00004b55cd736000, Names=ArrayRef<llvm::StringRef> @ 0x00000000806df988) at jitlayers.cpp:2154:22
    frame #3: 0x000000008e40f1e0 libjulia-codegen-debug.so.1.13`jl_compile_codeinst_now(codeinst=0x0000000000000000) at jitlayers.cpp:725:42
    frame #4: 0x000000008e40f7d0 libjulia-codegen-debug.so.1.13`jl_generate_fptr_for_unspecialized_impl(unspec=0x00004b55d1d03220) at jitlayers.cpp:924:13
    frame #5: 0x00000000869234b8 libjulia-internal-debug.so.1.13`jl_compile_method_internal(mi=0x00004b55d1c5d510, world=39) at gf.c:3551:13
    frame #6: 0x000000008692530c libjulia-internal-debug.so.1.13`_jl_invoke(F=0x0000000000000000, args=0x0000000000000000, nargs=0, mfunc=0x00004b55d1c5d510, world=39) at gf.c:4007:16
    frame #7: 0x0000000086925244 libjulia-internal-debug.so.1.13`ijl_invoke(F=0x0000000000000000, args=0x0000000000000000, nargs=0, mfunc=0x00004b55d1c5d510) at gf.c:4022:12
    frame #8: 0x000000008697d09c libjulia-internal-debug.so.1.13`jl_toplevel_eval_flex(m=0x00004b55d1d14010, e=0x00004b55d1c543b0, fast=1, expanded=1, toplevel_filename=0x00000000806e0a18, toplevel_lineno=0x00000000806e0a24) at toplevel.c:762:18
    frame #9: 0x000000008697ea04 libjulia-internal-debug.so.1.13`jl_parse_eval_all(module=0x00004b55d1d14010, text=0x00004b55cda8c100, filename=0x00004b55d1cd0190) at toplevel.c:890:22
    frame #10: 0x000000008697e570 libjulia-internal-debug.so.1.13`ijl_load_(module=0x00004b55d1d14010, filename=0x00004b55d1cd0190) at toplevel.c:934:26
    frame #11: 0x000000008697ebc8 libjulia-internal-debug.so.1.13`ijl_load(module=0x00004b55d1d14010, filename="boot.jl") at toplevel.c:947:26
    frame #12: 0x0000000086959310 libjulia-internal-debug.so.1.13`_finish_jl_init_(sysimage=(kind = JL_IMAGE_KIND_NONE, pointers = 0x0000000000000000, data = 0x0000000000000000, size = 0, base = 0), ptls=0x00004b55ccf49000, ct=0x00004b55cdd38010) at init.c:603:9
    frame #13: 0x0000000086958fd8 libjulia-internal-debug.so.1.13`ijl_init_(sysimage=(kind = JL_IMAGE_KIND_NONE, pointers = 0x0000000000000000, data = 0x0000000000000000, size = 0, base = 0)) at init.c:783:5
    frame #14: 0x00000000869afec0 libjulia-internal-debug.so.1.13`jl_repl_entrypoint(argc=5, argv=0x00000000806e0f60) at jlapi.c:1131:5
    frame #15: 0x00000000840f35c4 libjulia-debug.so.1.13`jl_load_repl(argc=15, argv=0x00000000806e0f60) at loader_lib.c:601:12
    frame #16: 0x0000000000210880 julia-debug`main(argc=15, argv=0x00000000806e0f60) at loader_exe.c:58:15
    frame #17: 0x0000000081baeca4 libc.so.7`__libc_start1(argc=15, argv=0x00000000806e0f60, env=0x00000000806e0fe0, cleanup=<unavailable>, mainX=(julia-debug`main at loader_exe.c:36)) at libc_start1.c:157:7
    frame #18: 0x00000000002107fc julia-debug`_start at crt1_s.S:62
```

This doesn't mean a whole lot to me but I'm hoping it does for somebody else who reads this. 😅

In case it's relevant, I'm using the default system compiler, which is Clang 18.1.5 on this version of FreeBSD. The most recent Julia commit I've attempted to build is fd521262bba028ac455c96df7de95db18ece8833 (2025-08-15), but I first observed this probably a week or so ago.

--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -8722,7 +8722,9 @@ static jl_llvm_functions_t
     if (src->inlining == 2)
         FnAttrs.addAttribute(Attribute::NoInline);
 
-#ifdef JL_DEBUG_BUILD
+    // Add strong stack protection for debug builds only when using the large code model,
+    // otherwise LLVM might try to relocate the stack canary out of range (see e.g. #59303)
+#if defined(JL_DEBUG_BUILD) && !(defined(_CPU_AARCH64_) || defined(_CPU_RISCV_))
     FnAttrs.addAttribute(Attribute::StackProtectStrong);
 #endif
 

